src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -885,7 +886,7 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
                         analyzedInitializers = InitializerRewriter.RewriteConstructor(processedInitializers.BoundInitializers, methodSymbol);
                         processedInitializers.HasErrors = processedInitializers.HasErrors || analyzedInitializers.HasAnyErrors;
 
-                        if (body != null && methodSymbol.ContainingType.IsStructType() && !methodSymbol.IsImplicitConstructor)
+                        if (body != null && ((methodSymbol.ContainingType.IsStructType() && !methodSymbol.IsImplicitConstructor) || instrumentForDynamicAnalysis))
```

Comment:
This cheap trick causes field initializers to be analyzed for each constructor that includes them, and so makes dynamic analysis work for field initializers. Because this is done only while compiling for instrumentation, there is a potential risk of different behavior when instrumentation is on vs. off. However, that seems better than throwing out the compiler's optimization of analyzing field initializers only once.